src/core/builtin_functions.cc

```diff
@@ -441,31 +441,42 @@ Value builtin_lookup(Arguments arguments, const Location& loc)
   double low_p, low_v, high_p, high_v;
   const auto& vec = arguments[1]->toVector();
 
-  // Second must be a vector of vec2, with valid numbers inside
+  // Second must be a vector of vec2,
   auto it = vec.begin();
-  if (vec.empty() || it->toVector().size() < 2 || !it->getVec2(low_p, low_v)) {
+  if (vec.empty() || it->toVector().size() < 2) {
     return Value::undefined.clone();
   }
-  high_p = low_p;
-  high_v = low_v;
-
-  for (++it; it != vec.end(); ++it) {
-    double this_p, this_v;
-    if (it->getVec2(this_p, this_v)) {
+  //first element of the vector must be a number
+  if (Value::Type::NUMBER  != it->toVector()[0].type()) {
+    return Value::undefined.clone();
+  }
+  //second element of the vector must be either a number or vector
+  if(Value::Type::NUMBER  == it->toVector()[1].type() || Value::Type::VECTOR == it->toVector()[1].type() ){
+    it->toVector()[0].getDouble(low_p);
+    high_p = low_p;
+    auto low_v  = it;
+    auto high_v = it;
+
+    for (++it; it != vec.end(); ++it) {
+      double this_p;
+      it->toVector()[0].getDouble(this_p);
+      
       if (this_p <= p && (this_p > low_p || low_p > p)) {
         low_p = this_p;
-        low_v = this_v;
+        low_v = it;
       }
       if (this_p >= p && (this_p < high_p || high_p < p)) {
         high_p = this_p;
-        high_v = this_v;
+        high_v = it;
       }
     }
+    if (p <= low_p) return high_v->toVector()[1].clone();
+    if (p >= high_p) return low_v->toVector()[1].clone();
+    double f = (p - low_p) / (high_p - low_p);
+    return (high_v->toVector()[1].operator*(f) + low_v->toVector()[1].operator*(1 - f));
```

Comment:
The first is more mathematically robust.  Namely, when f == 1.0, it will return exactly the "high" value.  
This is a very common "lerp" operation, and you can find lots of discussions and Q&As online where the first expression is preferred.

Reply:
OK